### PR TITLE
Refactoring of parts of event bus

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

**Comment by mbaechler** [comment]
```
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] [ERROR] inProcessingEventShouldBeReDispatchedToAnotherEventBusWhenOneIsDown  Time elapsed: 5.245 s  <<< ERROR!
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] com.jayway.awaitility.core.ConditionTimeoutException: Condition defined as a lambda expression in org.apache.james.mailbox.events.RabbitMQEventBusTest$AtLeastOnceTest that uses org.apache.james.mailbox.events.EventBusTestFixture$MailboxListenerCountingSuccessfulExecution expected:<[1]> but was:<[0]> within 5 seconds.
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] 	at org.apache.james.mailbox.events.RabbitMQEventBusTest$AtLeastOnceTest.inProcessingEventShouldBeReDispatchedToAnotherEventBusWhenOneIsDown(RabbitMQEventBusTest.java:213)
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] 
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] [INFO] 
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] [INFO] Results:
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] [INFO] 
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] [ERROR] Errors: 
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] [ERROR]   RabbitMQEventBusTest$AtLeastOnceTest.inProcessingEventShouldBeReDispatchedToAnotherEventBusWhenOneIsDown:213 » ConditionTimeout
[e717f8c381242c68ccfceb6fe3ccf64bd0f6d25f] [INFO] 
```

test this please

**Comment by chibenwa** [comment]
Can be considered :green_apple:

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/MailboxListenerRegistry.java [hunk 1]
```diff
@@ -19,41 +19,53 @@
...
     }
 
     void addListener(RegistrationKey registrationKey, MailboxListener listener, Runnable runIfEmpty) {
-        synchronized (listeners) {
-            if (listeners.get(registrationKey).isEmpty()) {
-                runIfEmpty.run();
-            }
-            listeners.put(registrationKey, listener);
-        }
+        listenersByKey.compute(registrationKey, (key, listeners) ->
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/MailboxListenerRegistry.java]: :+1: looks great... I always forgot `ConcurrentHashMap::compute`

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/MailboxListenerRegistry.java [hunk 2]
```diff
@@ -23,44 +23,77 @@
...
 import com.google.common.collect.ImmutableSet;
 import reactor.core.publisher.Flux;
 
 class MailboxListenerRegistry {
+
+    interface RemovalStatus {
+        boolean lastListenerRemoved();
+    }
+
+    public static class Registration {
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/MailboxListenerRegistry.java]: It looks confusing to me that we use the exact same wording than the EventBus API `Registration`.

What I do propose is to call this class `LocalRegistration`

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java [hunk 3]
```diff
@@ -89,11 +93,6 @@ void start() {
                 .doOnError(e -> structuredLogger(event, keys)
                     .log(logger -> logger.error("Exception happens when dispatching event", e)))
                 .onErrorResume(e -> Mono.empty()))
-            .cache()
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: It's done later on line 83

It do not break test and is simpler...

[the remark this answers is not in the document]
Comment (hoangdat) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: yep, I read the document. Thanks

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/MailboxListenerRegistry.java (hunk 2 above)
Comment (mbaechler) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/MailboxListenerRegistry.java]: I agree it is getting messy in term of naming. I think we should rename MailboxListenerRegistry in the first place.

**Comment by Arsnael** [comment]
Read it...

**Comment by mbaechler** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**